File: SubArray/SubarraySumEqualsK.java
import java.util.HashMap;

public class SubarraySumEqualsK {
    public static void main(String[] args) {
        int[] nums={1,2,3};
        int k=3;
        System.out.println(subarraySum(nums,k));
    }
    public static int subarraySum(int[] nums,int k){
        HashMap<Integer,Integer> mp=new HashMap<>();
        mp.put(0,1);
        int c=0;
        int sum=0;
        for(int i:nums){
            sum+=i;
            if(mp.containsKey(sum-k)){
                c+=mp.get(sum-k);
            }
            mp.put(sum,mp.getOrDefault(sum,0)+1);
        }
        return c;
    }
}

File: Pattern Wise DSA Sheet/Array/Prefix Sum/Prefix Sum + HashMap/SubarraySumEqualsK.py
# TC - O(n)
# SC - O(n)

class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        mp = {}
        mp[0] = 1

        s,c = 0, 0

        for i in range(len(nums)):
            s += nums[i]

            if s-k in mp:
                c += mp[s-k]

            mp[s] = mp.get(s, 0) + 1

        return c

        
